File: 199.Binary-Tree-Right-Side-View/tree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> list[int]:
        if not root:
            return []

        if not root.right:
            return [root.val] + self.rightSideView(root.left)

        return [root.val] + self.rightSideView(root.right)

File: 199.Binary-Tree-Right-Side-View/tree.rs
use std::{cell::RefCell, rc::Rc};

impl Solution {
    pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let Some(root) = root else {
            return Vec::new();
        };

        let mut root = root.borrow_mut();
        let l = root.left.take();
        let r = root.right.take();

        // O(n)
        let mut l = Self::right_side_view(l);
        let mut r = Self::right_side_view(r);
        r.extend(l.drain(r.len().min(l.len())..));

        let mut ret = vec![root.val];
        ret.append(&mut r);
        ret
    }
}
